src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -884,7 +884,24 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
         public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> source, int count)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            return SkipIterator<TSource>(source, count);
+
+            IList<TSource> sourceList = source as IList<TSource>;
+            return sourceList != null ? SkipList(sourceList, count) : SkipIterator<TSource>(source, count);
+        }
+
+        private static IEnumerable<TSource> SkipList<TSource>(IList<TSource> source, int count)
+        {
+            if (count < 0)
+            {
+                count = 0;
+            }
+
+            int sourceCount = source.Count;
```

Comment:
I know this contradicts what @stephentoub said in recommending this a commit ago, but all of the cases where calls to `Count` are pushed to a local are true locals and used to produce a result immediately rather than captured and used in an iterator. This means the only way it can race against something on the same thread is relatively obscure happening in a `Func`. In this case there could be something affecting `Count` within a `foreach` on the results of the skip. It might be more conservative to keep calling `Count`.